packages/ui-components/src/icon/jlicon.tsx

```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+import { classes } from 'typestyle';
+import { iconStyle, IIconStyle } from '../style/icon';
+
+export function createIcon(
```

Comment:
What about returning an instance of an `Icon` class that has two properties, `phosphor` and `react`, which return, respectively, a phosphor widget and a react element?

Reply:
AFAIK an instance cannot act as a `React.Component`, but I'm not 100% on that. Who on the jlab team right now is our React guru?

Reply:
Or did you just mean a subclass? That was my original design (you can see it in 27ac79517bd01619a6d06217c86319053fd4efe0), but there doesn't seem to a way to make it work with [ref forwarding](https://reactjs.org/docs/forwarding-refs.html) (which is a feature we definitely want in all of our more generic React components).

Reply:
Good point! What's the return type of `createIcon`? Is it possible to annotate? This looks like it works, but when I see https://github.com/jupyterlab/jupyterlab/pull/7299/files/b3a9e3756e3cd393afd034fb8b0837ee55ec9bdf#diff-a4bc5d52ceec5386efede6e47dc07d13R63-R65 I just wonder if we could have a class instead somewhere!

Reply:
It can't ever be a class, but I had attempted in d45419a to declare and export the return type of `createIcon`. Owing to some quirks in the way the `typeof` operator works I couldn't quite get it to work, but I think I know now how to work around the issue. In fact, I may be able to declare the return type of `createIcon` in a much simpler way, and avoid the use of `typeof` entirely. I'll take another stab at it

Reply:
React in general seems to be moving away from actual classes as fast as it possibly can, in favor of stuff like [hooks](https://reactjs.org/docs/hooks-intro.html). I don't know why they think classes and inheritance are so terrible

Reply:
`createIcon` now has a return type of `JLIcon.IComponent`, which is declared separately.